File: Task4/Kotlin/Task4_asyncIterator_2.kt
// Ongoing processing of large data sets that do not fit in memory

// with signal.cancel()

import kotlinx.coroutines.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlin.random.Random

fun <T> asyncIterator(array: List<T>): Flow<T> = flow {
    for (item in array) {
        delay(Random.nextLong(1000))
        emit(item)
    }
}

suspend fun <T, R> asyncMap(
    asyncIterable: Flow<T>,
    callback: suspend (T) -> R,
    signal: CompletableJob
): List<R> {
    val results = mutableListOf<R>()
    try {
        asyncIterable.collect { item ->
            if (!signal.isActive) {
                throw CancellationException("Aborted")
            }
            val result = callback(item)
            println("Processed: $result")
            results.add(callback(item))
        }
    } catch (e: CancellationException) {
        throw e
    }
    return results
}

fun demoFunc() = runBlocking {
    val controller = Job() // Аналог AbortController
    val signal = controller

    val numbers = listOf(1, 2, 3, 4, 5)
    val promiseTriple: suspend (Int) -> Int = { num -> num * 3 }

    try {
        val iterable = asyncIterator(numbers)

        launch {
            delay(1500)
            println("Aborting...")
            signal.cancel()
        }

        val results1 = asyncMap(iterable, promiseTriple, signal)
        println("Case 1 results: $results1")
    } catch (err: CancellationException) {
        println("Case 1 was aborted")
    } catch (err: Exception) {
        println("Error: ${err.message}")
    }
}

fun main() {
    demoFunc()
}

File: Task4/JavaScript/Task4_asyncIterator_2.js
// Ongoing processing of large data sets that do not fit in memory

// with controller.abort();

async function* asyncIterator(array) {
    for (const item of array) {
        yield new Promise((resolve) =>
            setTimeout(() => resolve(item), Math.random() * 1000)
        );
    }
}

async function asyncMap(asyncIterable, callback, signal) {
    const results = [];

    for await (const item of asyncIterable) {
        if (signal.aborted) {
            throw new DOMException("Aborted", "AbortError");
        }
        console.log(await callback(item))
        results.push(await callback(item));
    }

    return results;
}

async function demoFunc() {
    const controller = new AbortController();
    const { signal } = controller;

    const numbers = [1, 2, 3, 4, 5];
    const promiseTriple = (num) => Promise.resolve(num * 3);

    try {
        const iterable = asyncIterator(numbers);

        setTimeout(() => {
            console.log("Aborting...");
            controller.abort();
        }, 1500);

        const results1 = await asyncMap(iterable, promiseTriple, signal);
        console.log("Case 1 results:", results1);
    } catch (err) {
        if (err.name === "AbortError") {
            console.error("Case 1 was aborted");
        } else {
            console.error("Error:", err);
        }
    }
}

demoFunc()
    .then(() => console.log("demoFunc completed successfully"))
    .catch((err) => console.error("Error in demoFunc:", err));
